Issue 56426 a limit on the number of subcolumns in a table

garrettthomaskth opened the pull request:
<!--- Disable AI PR formatting assistant: true -->
### Changelog category (leave one):
- Improvement

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Add setting to limit number of subcolumns in persistent tables. Resolves #56426

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

### Details

https://github.com/ClickHouse/ClickHouse/issues/56426

Question: would it be better to put this check in the `validateTableStructure` function?
Question: Do I have to write my own documentation for the setting, or will docs automatically be generated based off the setting description?

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Introduces a new hard check during `CREATE/ATTACH TABLE` for disk-backed engines that can start rejecting schemas with many subcolumns; default limit (10000) may still impact edge cases. Changes are localized but affect DDL behavior and add a new error code.
> 
> **Overview**
> Adds a new setting, `max_static_subcolumns`, to cap the number of *static* subcolumns allowed when creating/attaching persistent (disk-backed) tables; `0` disables the check.
> 
> `InterpreterCreateQuery` now counts subcolumns from `ColumnsDescription` and throws the new `TOO_MANY_SUBCOLUMNS` error when the limit is exceeded, with a stateless SQL test covering both the disabled and failing cases. The setting is also recorded in `SettingsChangesHistory` and a new error code (`771`) is registered.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3e6147e955fae83e49c9e968c9955332a713d561. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

garrettthomaskth committed:
wip

garrettthomaskth committed:
check for storesDataOnDisk

garrettthomaskth committed:
add setting

garrettthomaskth committed:
add test

garrettthomaskth committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

garrettthomaskth committed:
expand on setting description

garrettthomaskth committed:
linter

garrettthomaskth committed:
add semi colons

garrettthomaskth committed:
fix duplicate subcolumn

garrettthomaskth committed:
fix order by id

garrettthomaskth committed:
make error message sound a little more natural

garrettthomaskth committed:
catch exception in test

garrettthomaskth committed:
remove error from reference file

Avogar commented on src/Core/Settings.cpp:
```diff
@@ -7037,6 +7037,9 @@ Possible values:
 - 0 — always,
 - negative integer - never.
+)", 0) \
+    DECLARE(UInt64, max_subcolumns, 10000, R"(
```
Unfortunately, this approach doesn't limit the number of dynamic subcolumns inside `JSON`/`Dynamic` types, because we determine dynamic subcolumns on insert when we have the data. So either we state that this limit doesn't include dynamic subcolumns, or we have to implement the check during insert into a table. We need to discuss what we want

Avogar replied:
New settings should be added to file `SettingsChangesHistory.cpp` where we keep track of changes in settings.

garrettthomaskth committed:
Apply suggestions from code review

Co-authored-by: Pavel Kruglov <48961922+Avogar@users.noreply.github.com>

garrettthomaskth committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

garrettthomaskth committed:
add getNumberOfSubcoumns function

garrettthomaskth commented:
> After internal discussion, we decided that it's ok to exclude dynamic subcolumns. But we need to add a note about it in the description of the setting

@Avogar should we change the name to `max_static_subcolumns` or `max_nondynamic_subcolumns` too?

garrettthomaskth committed:
add to settings history, disable check with negative value

garrettthomaskth committed:
fix negative spelling

garrettthomaskth committed:
max_subcolumns not unsigned

garrettthomaskth committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

garrettthomaskth committed:
remove accidental period

garrettthomaskth committed:
cast max_subcolumns to unsigned in comperison

garrettthomaskth committed:
make setting int64, rename setting

garrettthomaskth committed:
Rename test files. Fix unchanged setting name.

garrettthomaskth committed:
fix type

garrettthomaskth committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

garrettthomaskth committed:
Fix variable name in test.

Avogar commented on src/Core/Settings.cpp:
```diff
@@ -7291,6 +7291,9 @@ Possible values:
 - negative integer - never.
+)", 0) \
+    DECLARE(Int64, max_static_subcolumns, 10000, R"(
+Limit on the number of static subcolumns for persistent tables. Excludes dynamic columns in `JSON`, `Dynamic`, etc. types. Negative value disables check.
```
We usually use `0` as "unlimited" instead of negative values. Let's use it and change the type to `UInt64`.

garrettthomaskth committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

garrettthomaskth committed:
0 to bypass check

garrettthomaskth committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

Avogar commented on src/Interpreters/InterpreterCreateQuery.cpp:
```diff
@@ -138,6 +138,7 @@ namespace Setting
     extern const SettingsBool restore_replace_external_engines_to_null;
     extern const SettingsBool restore_replace_external_table_functions_to_null;
     extern const SettingsBool restore_replace_external_dictionary_source_to_null;
+    extern const SettingsInt64 max_static_subcolumns;
```
```suggestion
    extern const SettingsUInt64 max_static_subcolumns;
```

Avogar committed:
Apply suggestions from code review

Avogar committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

Avogar committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

garrettthomaskth committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

Avogar committed:
Merge branch 'master' of github.com:ClickHouse/ClickHouse into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

Avogar committed:
Respect max_static_subcolumns in ALTER commands

Avogar committed:
Address review comment

garrettthomaskth committed:
Merge remote-tracking branch 'clickhouse/master' into pr-89273

garrettthomaskth committed:
Only enforce max_static_subcolumns when ALTER increases the count

Address review comment
https://github.com/ClickHouse/ClickHouse/pull/89273#discussion_r3260560980:
unrelated ALTERs (`MODIFY COMMENT`, `MODIFY SETTING`, index/TTL alters,
`DROP COLUMN`) on a table that is already above the limit must not fail.
Compare the subcolumn count before vs after the ALTER and only reject
when the command itself pushes the count past the limit.

Extend `03645_max_static_subcolumns.sql` to cover an over-limit table:
unrelated ALTERs succeed, an `ADD COLUMN` that grows the count still
throws `TOO_MANY_SUBCOLUMNS`.

Avogar commented on src/Storages/AlterCommands.cpp:
```diff
@@ -1904,6 +1911,22 @@ void AlterCommands::validate(const StoragePtr & table, ContextPtr context) const
+    {
+        UInt64 max_static_subcolumns = context->getSettingsRef()[Setting::max_static_subcolumns];
+        size_t subcolumns_before = metadata->columns.getNumberOfSubcoumns();
+        size_t subcolumns_after = all_columns.getNumberOfSubcoumns();
```
It's a valid limitation, but this would be an unreal enge case in real world scenario I would say and I think "metadata copy" approach is overkill for the current scope

Avogar committed:
Udpate settings changes history

Avogar committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

Avogar committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

Avogar committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

Avogar committed:
Update SettingsChangesHistory.cpp

Avogar committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

Avogar committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

Avogar committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

Avogar committed:
Bump CI

Avogar committed:
Merge branch 'master' into issue-56426-a-limit-on-the-number-of-subcolumns-in-a-table

Avogar commented:
Something wrong went with the sync workflow, sync stopped creating. Let me close it and reopen a separate PR

Avogar commented:
https://github.com/ClickHouse/ClickHouse/pull/107733
